### PR TITLE
Refactors physiology into carbon, adds physio mods to keep track of mods and avoid duplicates

### PR DESCRIPTION
**Pull request by XDTM**
Should not have any gameplay change, besides fixing a potential interaction between nerve grounding and shock cookies that could grant permanent shock immunity.
Moved physiology into carbon since most of its functionality is relevant to carbon mobs.

Physiology is now modified with physio_mod datums; this way, duplicates of the same mod are discarded, as well as mods with the same id if the situation arises (for example, one could have a magic damage resistance effect at three different levels, and trying to apply more than one of them will only keep the one with the highest priority).

**Still untested!** Will test ASAP, opening it early for review

### CODE REVIEW

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 1]
```diff
@@ -1,3 +1,5 @@
+/mob/living/carbon/getarmor(def_zone, type)
+	return physiology.armor.getRating(d_type)
```
Comment (Qustinnus) [inline code/modules/mob/living/carbon/carbon_defense.dm]: d_type is undefined

File: code/modules/mob/living/carbon/physiology.dm [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+	
+/datum/physiology/proc/apply_mod(datum/physio_mod/mod_type)
+	for(var/i in physio_mods)
+		var/datum/physio_mod/P = i
+		if(P.id == initial(mod_type.id))
+			if(P.priority <= initial(mod_type.priority))
+				remove_mod_id(P.id)
+			else
+				return
+	var/new_mod = new mod_type
```
Comment (Qustinnus) [inline code/modules/mob/living/carbon/physiology.dm]: you need to type the var

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by Razharas** [comment]
Id like to point out that while this is a fine way of handling var clutter is not a real solution of deentangling the mob code
True way is clustering specific features with specific objects they are supposed to represent, so health mods should be in limbs on per-limb bases, brain damage in brain and so on

**Comment by XDTM** [comment]
I actually agree and still think this code is really imprefect; what i'm trying to achieve is more or less a generalized system to handle all these numeric vars that are modified by multiple sources, sorta like traits do for bools.

Technically all these vars could be spread across their owners, but you'd go back to having issues with multiple sources (i.e. spell gives you 0.5 brutemod, then doubles it when it's over; however if you replace a limb in the meantime you end up with a limb with permanent 2 brutemod instead of 1).

Perhaps it could be componentized somehow, but it would need some planning (plus sorting out what component are and are not supposed to do)

**Comment by Razharas** [comment]
Well it just would need a trait-tracker on per-limb per-organ basis, instead of per-mob

**Comment by XDTM** [comment]
Some stuff can only be applied on the mob itself, though, while some other stuff does not focus a particular limb but affects a modifier that would be limb based, like generic brute resistance. In that case you'd have a status effect that edits all the limbs trait trackers, and you would have the same problem with limb replacement.

I'm currently reworking this into a system that updates itself on every change, so it resets to start and applies all active mods all over again. Shouldn't be too expensive, and prevents all those tricky issues. The major positive is that it will allow mixing addition and multiplication without messing up the reverse operation.

**Comment by optimumtact** [comment]
these are essentially armour via another name, they should just be added to the armour handling

**Comment by XDTM** [comment]
merging armour and those vars actually sounds like a good idea, although some vars like movespeed don't really fit in the armor context

**Comment by Razharas** [comment]
Instead of "adding" it to armour you can "merge" armour and this to get a system that handles such things in general

**Pull request closed without merging**